Classes/Service/NodeService.php

```diff
@@ -0,0 +1,219 @@
+        $queryBuilder = $this->createQueryBuilder($workspaceChain);
+        $queryBuilder->andWhere('n.nodeType IN (:includeNodeTypes)');
+        $queryBuilder->setParameter('includeNodeTypes', $this->getNodeTypeFilter($configurationDto));
+        $iterator = $queryBuilder->getQuery()->iterate();
+
+        $nodeDatasThatNeedProcessing = [];
+        $nodeDatasThatNeedProcessingCount = 0;
+        $iteratedItems = 0;
+
+        /** @var NodeData $nodeData */
+        foreach ($this->iterate($iterator) as $nodeData) {
+            $iteratedItems++;
+            if ($iteratedItems <= $configurationDto->getFirstResult()) {
+                continue;
+            }
+
+            if ($nodeDatasThatNeedProcessingCount >= $configurationDto->getLimit()) {
+                break;
+            }
+
+            $currentFocusKeywordPropertyValue = $nodeData->hasProperty('focusKeyword') ? $nodeData->getProperty('focusKeyword') : null;
+            if (!self::matchFocusKeywordProperty($currentFocusKeywordPropertyValue, $configurationDto)) {
+                continue;
+            }
+
+            $nodeDatasThatNeedProcessing[] = $nodeData;
+            $nodeDatasThatNeedProcessingCount++;
+        }
+
+        $nodeDatasThatNeedProcessing = $this->reduceNodeVariantsByWorkspaces($nodeDatasThatNeedProcessing, $workspaceChain);
+
+        $nodesThatNeedProcessing = [];
+        foreach (array_slice($nodeDatasThatNeedProcessing, $configurationDto->getFirstResult(), $configurationDto->getLimit()) as $nodeData) {
+            $context = $this->createContentContext($configurationDto->getWorkspace(), $nodeData->getDimensionValues());
+            $node = new Node($nodeData, $context);
+            $nodesThatNeedProcessing[] = $this->focusKeywordModuleResultDtoFactory->createFromNode($node, $controllerContext);
+        }
+
+        return new ResultCollectionDto(
+            $nodesThatNeedProcessing,
+            $configurationDto->getFirstResult() + sizeof($nodesThatNeedProcessing)
+        );
+    }
+
+    /**
+     * @param array<FocusKeywordModuleResultDto> $resultDtos
+     *
+     * @return void
+     */
+    public function updateFocusKeywordOnNodes(array $resultDtos): void
+    {
+        foreach($resultDtos as $resultDto) {
+            /** @var array{nodePath: string, workspaceName: string, dimensions: array} $contextPathSegments */
+            $contextPathSegments = NodePaths::explodeContextPath($resultDto->getNodeContextPath());
+            $context = $this->createContentContext($contextPathSegments['workspaceName'],
+                $contextPathSegments['dimensions']);
+            $node = $context->getNode($contextPathSegments['nodePath']);
+            $node->setProperty('focusKeyword', $resultDto->getFocusKeyword());
+        }
+    }
+
+    protected static function matchFocusKeywordProperty(mixed $value, FocusKeywordModuleConfigurationDto $configurationDto): bool
+    {
+        if ($configurationDto->getMode() === 'both') {
```

Comment:
Das wär ein klassiches Switch Statemeng, um klar zu machen das basierend aus dem Mode die Auswahl passiert:

```

        switch ($configurationDto->getMode()) {
            case 'both':
                return true;
            case 'only-empty':
                return empty($value);
            case 'only-existing':
                return !empty($value);
        }
```

Reply:
PHPStorm hat mir sogar das angeboten:

```
return match ($configurationDto->getMode()) {
    'both' => true,
    'only-empty' => empty($value),
    'only-existing' => !empty($value),
    default => false,
};
```

Reply:
`match` gibt's allerdings erst mit PHP 8..

Reply:
Finde ich okay, PHP 7 is eh nimma sicher https://www.php.net/supported-versions.php